stdlib/immutablemap.gr

```diff
@@ -0,0 +1,495 @@
+/**
+ * @module ImmutableMap: An ImmutableMap holds key-value pairs. Any value may be used as a key or value. Operations on an ImmutableMap do not mutate the map's internal state.
+ * @example import ImmutableMap from "immutablemap"
+ *
+ * @since v0.5.4
+ */
+
+import List from "list"
+import Array from "array"
+import Option from "option"
+
+// implementation based on the paper "Implementing Sets Efficiently in a
+// Functional Language" by Stephen Adams
+
+record Node<k, v> {
+  key: k,
+  val: v,
+  size: Number,
+  left: ImmutableMap<k, v>,
+  right: ImmutableMap<k, v>,
+},
+enum ImmutableMap<k, v> {
+  Empty,
+  Tree(Node<k, v>),
+}
+
+/**
+ * @section Values: Functions for working with ImmutableMaps.
+ */
+
+// semi-arbitrary value chosen for algorithm for determining when to balance
+// trees; no tree can have a left subtree containing this number of times
+// more elements than its right subtree or vice versa
+let weight = 4
```

Comment:
If something like this were to be done then should the regular `make` function still be removed?

Reply:
I think `make` should always be removed because we would want to be explicit with the name, such as `makeWeighted`. If we ever get optional arguments, that choice might change.